src/irmin/contents.ml

```diff
@@ -32,6 +32,139 @@ module Cstruct = struct
+  |~ case1 "array" (list ty) (fun arr -> `A arr)
+  |> sealv)
+
+module Json = struct
+  type t = (string * json) list
+
+  let t = Type.(list (pair string json))
+
+  let merge_object a b =
+    try
+      List.fold_right (fun (k, v) acc ->
+        match List.assoc_opt k b with
+        | Some x when v <> x -> failwith "Unable to merge JSON objects"
+        | Some _ -> acc
+        | _ -> (k, v) :: acc
+      ) a b
+      |> Merge.ok
+    with
+      Failure msg -> Merge.conflict "%s" msg
+
+  let merge ~old t1 t2 =
+    let open Merge.Infix in
+    old () >>=* function
+      | Some j ->
+          if j = t1 then
+            merge_object t1 t2
+          else if j = t2 then
+            merge_object t2 t1
+          else
+            merge_object j t1 >>=* fun t1 ->
+            merge_object t1 t2
+      | None -> merge_object t1 t2
+
+  let merge = Merge.(option (v t merge))
+
+  let lexeme e x = ignore (Jsonm.encode e (`Lexeme x))
+
+  let rec encode_json e = function
+    | `Null -> lexeme e `Null
+    | `Bool b -> lexeme e (`Bool b)
+    | `String s -> lexeme e (`String s)
+    | `Float f -> lexeme e (`Float f)
+    | `A a ->
+        lexeme e `As;
+        List.iter (encode_json e) a;
+        lexeme e `Ae;
+    | `O o ->
+        lexeme e `Os;
+        List.iter (fun (k, v) ->
+          lexeme e (`Name k);
+          encode_json e v
+        ) o;
+        lexeme e `Oe
+
+  let pp fmt x =
+    let buffer = Buffer.create 32 in
+    let encoder = Jsonm.encoder (`Buffer buffer) in
+    encode_json encoder (`O x);
+    ignore @@ Jsonm.encode encoder `End;
+    let s = Buffer.contents buffer in
+    Fmt.pf fmt "%s" s
+
+  let decode_json d =
+    let decode d = match Jsonm.decode d with
+      | `Lexeme l -> l
+      | `Error e -> failwith (Fmt.strf "%a" Jsonm.pp_error e)
+      | _ -> failwith "invalid JSON encoding"
+    in
+    let rec unwrap v d = match v with
+      | `Os -> obj [] d
+      | `As -> arr [] d
+      | `Null | `Bool _ | `String _ | `Float _ as v -> v
+      | _ -> failwith "invalid JSON value"
+    and arr vs d =
+      match decode d with
+      | `Ae -> `A (List.rev vs)
+      | v ->
+          let v = unwrap v d in
+          arr (v::vs) d
+    and obj ms d =
+      match decode d with
+      | `Oe -> `O (List.rev ms)
+      | `Name k ->
+          let v = unwrap (decode d) d in
+          obj ((k, v) :: ms) d
+      | _ -> failwith "invalid JSON object"
+    in
+    try
+      Ok (unwrap (decode d) d)
+    with
+      | Failure msg -> Error (`Msg msg)
+
+  let of_string s =
+    let decoder = Jsonm.decoder (`String s) in
+    match decode_json decoder with
+    | Ok (`O obj) -> Ok obj
+    | Ok _ -> Error (`Msg "Irmin JSON values must be objects")
+    | Error _ as err -> err
+
+  module Make(S: S.S0) = struct
+    include S
+    let pp = Type.pp_json t
+    let of_string s =
+      let decoder = Jsonm.decoder (`String s) in
+      Type.decode_json t decoder
+    let merge = Merge.(option (idempotent t))
```

Comment:
I am not sure to fully understand the purpose of `Make`, but I guess you probably want to re-use the `merge` function defined above here  instead of using the default one?

Reply:
`Make` was just added as a clear way to make a JSON value from an exsiting type, rather than the user having to combine all those pieces manually. It is not entirely necessary, but I found it helpful when testing these changes.

Reply:
Despite it being convenient for my testing, I'm not sure that it will work because the merge function would have to be rewritten for this. I am going to remove it, but add it as an example to one of the tutorials.